File: lib/cadence_web/live/ops_console_live/widgets/value_display_widget.ex
defmodule CadenceWeb.OpsConsoleLive.Widgets.ValueDisplayWidget do
  @moduledoc """
  Phoenix LiveComponent for displaying a single telemetry value.

  Shows the current value with:
  - Limits state coloring (green/yellow/red)
  - Trend indicator (up/down/stable)
  - Unit display
  - Item label

  Real-time updates are handled by a JavaScript hook that subscribes
  to the TelemetryStore for high-frequency updates without LiveView
  round-trips.

  ## Usage

      <.live_component
        module={ValueDisplayWidget}
        id="value-widget-1"
        widget_id="widget-123"
        config={%{
          "telemetry_item" => %{
            "target" => "SAT-1",
            "packet" => "HEALTH",
            "item" => "CPU_TEMP"
          },
          "unit" => "°C",
          "precision" => 2,
          "show_trend" => true
        }}
        on_configure={JS.push("open_widget_config")}
      />
  """
  use CadenceWeb, :live_component

  import CadenceWeb.WidgetComponents

  alias Phoenix.LiveView.JS

  @impl true
  def mount(socket) do
    {:ok, assign(socket, current_value: nil, limits_state: :unknown, trend: nil)}
  end

  @impl true
  def update(assigns, socket) do
    socket =
      socket
      |> assign(assigns)
      |> assign_telemetry_key()

    {:ok, socket}
  end

  defp assign_telemetry_key(socket) do
    config = socket.assigns[:config] || %{}
    item = config["telemetry_item"] || %{}

    telemetry_key =
      case {item["target"], item["packet"], item["item"]} do
        {t, p, i} when is_binary(t) and is_binary(p) and is_binary(i) ->
          "#{t}:#{p}:#{i}"

        _ ->
          nil
      end

    assign(socket, :telemetry_key, telemetry_key)
  end

  @impl true
  def render(assigns) do
    config = assigns[:config] || %{}
    item = config["telemetry_item"] || %{}
    unit = config["unit"] || ""
    show_trend = Map.get(config, "show_trend", true)
    title = config["title"] || item["item"] || "Value"

    assigns =
      assigns
      |> assign(:unit, unit)
      |> assign(:show_trend, show_trend)
      |> assign(:title, title)
      |> assign(:item, item)

    ~H"""
    <.widget
      id={@widget_id}
      title={@title}
      on_configure={@on_configure && JS.push("open_widget_config", value: widget_config_value(@widget_id, "value_display", @config))}
    >
      <div
        id={"#{@widget_id}-content"}
        class="value-display"
        phx-hook="ValueDisplayWidget"
        phx-update="ignore"
        data-telemetry-key={@telemetry_key}
        data-unit={@unit}
        data-precision={@config["precision"] || 2}
        data-show-trend={to_string(@show_trend)}
      >
        <div class="value-container">
          <span class="value text-4xl font-mono font-bold" data-value>--</span>
          <span :if={@unit != ""} class="unit text-lg text-base-content/70">{@unit}</span>
          <span :if={@show_trend} class="trend text-2xl" data-trend></span>
        </div>
        <div class="limits-state mt-2" data-limits-state>
          <.limits_badge state={:unknown} />
        </div>
        <div :if={@item["target"]} class="item-label text-sm text-base-content/50 mt-2">
          {@item["target"]}/{@item["packet"]}/{@item["item"]}
        </div>
      </div>
    </.widget>
    """
  end

  defp widget_config_value(widget_id, widget_type, config) do
    %{
      widget_id: widget_id,
      widget_type: widget_type,
      config: config
    }
  end
end

File: legacy/cadence_legacy/assets/js/widgets/value_display_widget.js
/**
 * Value Display Widget
 *
 * Displays the current value of a single telemetry item with limits coloring.
 * Shows value, unit, limits state, and optional trend indicator.
 */

import { BaseWidget, registerWidget } from "../dashboard/widget_factory"
import { getStore } from "../telemetry_store"
import { getLimitsColor, TokyoNightColors } from "../scichart/tokyo_night_theme"

/**
 * ValueDisplayWidget - Shows current value with limits coloring.
 */
export class ValueDisplayWidget extends BaseWidget {
  static displayName = "Value Display"
  static description = "Shows current telemetry value with limits status"
  static icon = "number"

  constructor(options) {
    super(options)
    this.type = "value_display"
    this.currentValue = null
    this.limitsState = "unknown"
    this.previousValue = null
    this.trend = null  // "up" | "down" | "stable"
    this.valueElement = null
    this.trendElement = null
    this.stateElement = null
  }

  init() {
    this.render()
    this._subscribeToTelemetry()
    this.initialized = true
  }

  renderContent() {
    const item = this.config.telemetry_item || {}
    const unit = this.config.unit || ""

    return `
      <div class="value-display flex flex-col items-center justify-center h-full">
        <div class="value-container flex items-baseline gap-2">
          <span class="value text-4xl font-mono font-bold" id="value-${this.id}">--</span>
          <span class="unit text-lg text-base-content/70">${unit}</span>
          <span class="trend text-2xl" id="trend-${this.id}"></span>
        </div>
        <div class="limits-state mt-2" id="state-${this.id}">
          <span class="badge badge-sm badge-neutral">Unknown</span>
        </div>
        <div class="item-label text-sm text-base-content/50 mt-2">
          ${item.target || ""}/${item.packet || ""}/${item.item || "Item"}
        </div>
      </div>
    `
  }

  render() {
    super.render()

    // Get references to dynamic elements
    this.valueElement = this.container.querySelector(`#value-${this.id}`)
    this.trendElement = this.container.querySelector(`#trend-${this.id}`)
    this.stateElement = this.container.querySelector(`#state-${this.id}`)
  }

  _subscribeToTelemetry() {
    const item = this.config.telemetry_item

    if (!item || !item.target || !item.packet || !item.item) {
      console.warn("[ValueDisplayWidget] No telemetry item configured")
      return
    }

    try {
      const store = getStore()
      const key = `${item.target}:${item.packet}:${item.item}`

      // Get current value if available
      const current = store.getCurrent(key)
      if (current) {
        this._updateDisplay(current.value, current.limitsState)
      }

      // Subscribe to updates
      const unsub = store.subscribe(key, (value, timestamp, limitsState) => {
        this._updateDisplay(value, limitsState)
      })

      this.subscriptions.push(unsub)
    } catch (err) {
      console.error("[ValueDisplayWidget] Failed to subscribe to telemetry:", err)
    }
  }

  _updateDisplay(value, limitsState) {
    // Calculate trend
    if (this.currentValue !== null && typeof value === "number") {
      if (value > this.currentValue) {
        this.trend = "up"
      } else if (value < this.currentValue) {
        this.trend = "down"
      } else {
        this.trend = "stable"
      }
    }

    this.previousValue = this.currentValue
    this.currentValue = value
    this.limitsState = limitsState

    // Update value display
    if (this.valueElement) {
      this.valueElement.textContent = this._formatValue(value)

      // Apply limits color
      const color = getLimitsColor(limitsState)
      this.valueElement.style.color = color
    }

    // Update trend indicator
    if (this.trendElement && this.config.show_trend !== false) {
      const trendIcons = {
        up: "&#x2191;",    // ↑
        down: "&#x2193;",  // ↓
        stable: "&#x2022;" // •
      }
      const trendColors = {
        up: TokyoNightColors.teal,
        down: TokyoNightColors.red,
        stable: TokyoNightColors.textMuted
      }

      this.trendElement.innerHTML = trendIcons[this.trend] || ""
      this.trendElement.style.color = trendColors[this.trend] || ""
    }

    // Update limits state badge
    if (this.stateElement) {
      const badgeClass = this._getLimitsBadgeClass(limitsState)
      this.stateElement.innerHTML = `
        <span class="badge badge-sm ${badgeClass}">${this._formatLimitsState(limitsState)}</span>
      `
    }
  }

  _formatValue(value) {
    if (value === null || value === undefined) return "--"

    const precision = this.config.precision ?? 2

    if (typeof value === "number") {
      if (Number.isInteger(value)) {
        return value.toString()
      }
      return value.toFixed(precision)
    }

    return String(value)
  }

  _formatLimitsState(state) {
    const labels = {
      green: "Normal",
      yellow: "Warning",
      yellow_low: "Low Warning",
      yellow_high: "High Warning",
      red: "Critical",
      red_low: "Low Critical",
      red_high: "High Critical",
      blue: "Stale",
      unknown: "Unknown"
    }
    return labels[state] || "Unknown"
  }

  _getLimitsBadgeClass(state) {
    const classes = {
      green: "badge-success",
      yellow: "badge-warning",
      yellow_low: "badge-warning",
      yellow_high: "badge-warning",
      red: "badge-error",
      red_low: "badge-error",
      red_high: "badge-error",
      blue: "badge-info",
      unknown: "badge-neutral"
    }
    return classes[state] || "badge-neutral"
  }

  openConfig() {
    const event = new CustomEvent("widget:configure", {
      detail: {
        widgetId: this.id,
        widgetType: this.type,
        config: this.getConfig()
      },
      bubbles: true
    })
    this.container.dispatchEvent(event)
  }
}

// Register the widget
registerWidget("value_display", ValueDisplayWidget)

export default ValueDisplayWidget
